### PR TITLE
Support for TypeScript by providing typings files

### LINKED ISSUE #589: TypeScript Support [Help Wanted]
TypeScript support is made up 100% from community contributions. None of the core maintainers know or use TypeScript and thus are unable to offer support for it. We believe that TypeScript has value to some and are willing to merge in changes the community find valuable. 

If you would like to make Middy better, please consider submitting, reviewing, and/or commenting on PRs/issues. We generally like to give a week before merging a PR to give the community time to comment.

[TypeScript Issues](https://github.com/middyjs/middy/issues?q=is%3Aopen+is%3Aissue+label%3Atype-systems)

### PR DESCRIPTION
**Pull request by lmammino**
This PR explores an approach to provide typings for v2 and support TypeScript users.

This work is totally experimental and currently incomplete. It's just a starting point to validate ideas and collect feedback.

Right now I am trying to figure out how the typescript experience would look like and using [`tsd`](https://npm.im/tsd) to validate locally that the typings make sense.

The current status is that I partially wrote typings only for the `core` package.

A possible usage might look like this:

```typescript
import middy from '@middy/core'
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda'

async function baseHandler (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
  // the returned response will be checked against the type `APIGatewayProxyResult`
  return {
    statusCode: 200,
    body: `Hello from ${event.path}`
  }
}

// handler is type middy.MiddyfiedHandler<APIGatewayProxyEvent, APIGatewayProxyResult, Error>
let handler = middy(baseHandler)


type Request = middy.Request<APIGatewayProxyEvent, APIGatewayProxyResult, Error>
let myCustomMiddleware = {
  before: (request: Request) => {
    console.log('Before', request)
  },
  after: (request: Request) => {
    console.log('After', request)
  },
  onError: (request: Request) => {
    console.log('OnError', request)
  }
}

handler
  .use(myCustomMiddleware)

export default handler
```

What I like from this example is that:

 - You get nice autocomplete for the event fields.
 - You also get validation if your response does not comply with the response type (e.g. if you are not providing a statusCode -number- and a body -string-)

## TODO

- [x] Decide on whether we like this approach or not
- [x] Update branch with the latest changes in `release/2.x`
- [x] Remove `standard` and switch to full ESLint (using standard + typescript as presets)
- [x] Add `tsd` tests as part of the test suite
- [x] Add manual typings for all the packages
  - [x] `core`
  - [x] `do-not-wait-for-empty-event-loop`
  - [x] `error-logger`
  - [x] `http-content-negotiation`
  - [x] `http-cors`
  - [x] `http-error-handler`
  - [x] `http-event-normalizer`
  - [x] `http-header-normalizer`
  - [x] `http-json-body-parser`
  - [x] `http-multipart-body-parser`
  - [x] `http-partial-response`
  - [x] `http-response-serializer`
  - [x] `http-security-headers`
  - [x] `http-urlencode-body-parser`
  - [x] `http-urlencode-path-parser`
  - [x] `input-output-logger`
  - [x] `rds-signer`
  - [x] `s3-key-normalizer`
  - [x] `secrets-manager`
  - [x] `sqs-json-body-parser`
  - [x] `sqs-partial-batch-failure`
  - [x] `ssm`
  - [x] `sts`
  - [x] `util`
  - [x] `validator`
  - [x] `cloudwatch-metrics`
- [x] Make tests green
- [x] Add "Use with TypeScript" section in the README

Closes #589

### CODE REVIEW

File: packages/core/index.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    console.log('After', handler)
+  },
+  onError: (handler: Handler) => {
+    console.log('OnError', handler)
+  },
+}])
+expectType<Handler>(handler)
+
+// before
+handler.before(function (handler: Handler) {
```
Comment (lmammino) [inline packages/core/index.test-d.ts:48]: This one currently fails for some mismatch of types... got stuck here trying to figure out why...

```
Argument of type '(handler: Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>) => void' is not assignable to parameter of type 'Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>'.
  Type '(handler: Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>) => void' is not assignable to type 'Handler<APIGatewayProxyEvent, APIGatewayProxyResult>'.
    Types of parameters 'handler' and 'event' are incompatible.
      Type 'APIGatewayProxyEvent' is not assignable to type 'Middy<APIGatewayProxyEvent, APIGatewayProxyResult, Error>'.
        Type 'APIGatewayProxyEventBase<{ [name: string]: any; }>' is not assignable to type 'Handler<APIGatewayProxyEvent, APIGatewayProxyResult>'.
          Type 'APIGatewayProxyEventBase<{ [name: string]: any; }>' provides no match for the signature '(event: APIGatewayProxyEvent, context: Context, callback: Callback<APIGatewayProxyResult>): void | Promise<APIGatewayProxyResult>'.
```

That message sounds a bit cryptic to me right now...

Comment (lmammino) [reply packages/core/index.test-d.ts]: I figured this one out. I did not account that we need a `MiddlewareAttachFn` and a `MiddlewareFn` as 2 distinct concepts! :D

File: packages/core/index.d.ts [hunk 2]
```diff
@@ -1,57 +1,72 @@
...
-  T extends (event: any, context: C, callback: Callback<infer RetType>) => void ? RetType :
-  never;
+declare type PluginObject = {
+  beforePrefetch?: PluginHook;
+  requestStart?: PluginHook;
+  beforeMiddleware?: PluginHookWithMiddlewareName;
+  afterMiddleware?: PluginHookWithMiddlewareName;
+  beforeHandler?: PluginHook;
+  afterHandler?: PluginHook;
+  requestEnd?: PluginHook;
```
Comment (willfarrell) [inline packages/core/index.d.ts:16]: This can be a promise, not sure if that needs a different type

File: packages/core/index.js [hunk 3]
```diff
@@ -1,78 +1,4 @@
...
- * @param {function} handler - the original handler function.
- *   It will expose properties `event`, `context`, `response`, `error` and `internal` that can
- *   be used to interact with the middleware lifecycle
- * @return {void|Promise} - A middleware can return a Promise.
- *                          In this case middy will wait for the promise to resolve (or reject) and it will automatically
- *                          propagate the result to the next middleware.
- */
-
-const middy = (handler = () => {}, plugin) => {
+const middy = (handler = () => { }, plugin = null) => {
```
Comment (willfarrell) [inline packages/core/index.js:1]: any reason you're casting plugin to `null`? undefined is considered nullish

Comment (lmammino) [reply packages/core/index.js:1]: I was trying to make it optional... Maybe i should revert that!

File: packages/core/index.d.ts (hunk 2 above)
Comment (lmammino) [reply packages/core/index.d.ts:16]: Great point, yeah we will need the method to return a promise...

File: packages/do-not-wait-for-empty-event-loop/package.json [hunk 4]
```diff
@@ -44,5 +43,9 @@
     "url": "https://github.com/middyjs/middy/issues"
   },
   "homepage": "https://github.com/middyjs/middy#readme",
-  "gitHead": "7a6c0fbb8ab71d6a2171e678697de9f237568431"
+  "gitHead": "7a6c0fbb8ab71d6a2171e678697de9f237568431",
+  "devDependencies": {
+    "@middy/core": "2.0.0-alpha.6",
+    "@types/aws-lambda": "^8.10.3"
```
Comment (willfarrell) [inline packages/do-not-wait-for-empty-event-loop/package.json:49]: is this needed? it's included in @middy/core

Comment (lmammino) [reply packages/do-not-wait-for-empty-event-loop/package.json:49]: It's included there as a dev dependency so we need it here as well I believe

Comment (willfarrell) [reply packages/do-not-wait-for-empty-event-loop/package.json:49]: oh right, it need to ref the lambda context

**Review by willfarrell — APPROVED** [review]

**Comment by willfarrell** [comment]
final comments:
- do we need `tsconfig.json`? can it be merged into package.json? Not a big deal.
- can the `index.test-d.ts` be move the tests to `__tests__/index.d.ts`?

**Comment by lmammino** [comment]
@willfarrell 

> do we need `tsconfig.json`? can it be merged into package.json? Not a big deal.

It should be possible, but I tried that and doesn't seem to be working well with ts-standard :(

>  can the `index.test-d.ts` be moved to `__tests__/index.d.ts`?

I tried to make this work but I don't think it's something configurable, unfortunately...

**Comment by lmammino** [comment]
@willfarrell scrap the last one... maybe i found a way to make it work! give me 5 mins!

**Comment by lmammino** [comment]
@willfarrell 

Nope, the `tsd` test file needs to be in the same folder as the typing definitions (`index.d.ts`) and I think that makes more sense in the root folder of the package...

**Pull request merged**